stripe.go

```diff
@@ -621,11 +631,37 @@ func (s *BackendImplementation) shouldRetry(err error, req *http.Request, resp *
 
 	stripeErr, _ := err.(*Error)
 
-	// TODO: This retries any non-Stripe errors produced as part of the
-	// communication process. It generally works, but there are many errors
-	// that should *not* be retried. Try to make this more granular by
-	// including only connection errors, timeout errors, etc.
+	// Don't retry if the context was was cancelled or its deadline was
+	// exceeded.
+	if req.Context() != nil && req.Context().Err() != nil {
```

Comment:
req.Context() being nil or it's error not being nil means it's deadline was exceeded or cancelled. How do we know this? Can we add an end to end test (https://github.com/stripe/stripe-go/pull/1187/files#diff-1c7223ac1329b65f54265bc927d22a78R68) where the Context is explicitly cancelled?

Reply:
@ctrudeau-stripe Basically going by the documentation on this one. See [context.Context](https://godoc.org/context#Context):

``` go
    // If Done is not yet closed, Err returns nil.
    // If Done is closed, Err returns a non-nil error explaining why:
    // Canceled if the context was canceled
    // or DeadlineExceeded if the context's deadline passed.
    // After Err returns a non-nil error, successive calls to Err return the same error.
    Err() error
```

Don't you think the existing test is good enough? It does basically what you requested: generates a context, explicitly cancels it, and then check that we bailed out of this function with `false`.

Reply:
Ah that sounds great to me.

Reply:
@ctrudeau-stripe Awesome. Okay if I call this PR approved then? I'll write a follow up with more detailed logging after.

Reply:
:shipit: LGTM